axiom-eth/src/rlp/mod.rs

```diff
@@ -625,4 +807,268 @@ impl<'range, F: ScalarField> RlpChip<'range, F> {
+                ctx,
+                &rlp_array,
+                &len_start_id,
+                len_len.value(),
+                max_item_len_len,
+            );   // ***** unconstrained
+
+            // 0 if phantom, literal or short
+            // len if long
+            let len_byte_val = evaluate_byte_array(ctx, self.gate(), &len_cells, len_len);
+                // constrained
+
+            // len if short or long
+            // 1 if phantom or literal
+            let mut item_len = self.gate().select(
+                ctx,
+                len_byte_val,
+                prefix_parsed.next_len,
+                prefix_parsed.is_big,
+            );
+
+
+            // prefix_len is 0 if phantom or literal
+            // 1 otherwise
+            let mut prefix_len = prefix_parsed.is_not_literal;
+
+            if is_variable_len{
+                // If `prefix_idx >= rlp_len`, that means we are done
+                // Question: can we change this to an equality check?
+                let not_phantom = self.range.is_less_than(
+                    ctx,
+                    prefix_idx,
+                    rlp_len,
+                    bit_length(max_rlp_array_len as u64),
+                );
+
+
+                // In cases where the RLP sequence is a list of unknown variable length, we keep track
+                // of whether the corresponding index actually is a list item by constraining that
+                // all of `prefix_len, len_len, field_len` are 0 when the current field should be treated
+                // as a phantom and not actually in the list
+                prefix_len = self.gate().mul(ctx, prefix_len, not_phantom);
+                len_len = self.gate().mul(ctx, len_len, not_phantom);
+
+                // len if short or long
+                // 1 if literal
+                // 0 if phantom
+                item_len = self.gate().select(
+                    ctx,
+                    item_len,
+                    Constant(F::zero()),
+                    not_phantom,
+                );
+            }
+
+            // this is just print for debug
+            let mut prefix_idx_val = F::from(0u64);
+            let mut item_len_val = F::from(0u64);
+
+            prefix_idx_val = *prefix_idx.value();
+            item_len_val = *item_len.value();
+            println!("This item starts at {:?} and has length {:?}", prefix_idx_val.get_lower_32(), item_len_val.get_lower_32());
+            // end print for debug
+
+
+
+            // item_len is 0 if phantom, 1 if literal, ...
+            self.range.check_less_than_safe(ctx, item_len, (max_item_len + 1) as u64);
+
+            let item_cells = witness_subarray(
+                ctx,
+                &rlp_array,
+                &(len_start_id + len_len.value()),
+                item_len.value(),
+                max_item_len,
+            ); // *** unconstrained
+            running_max_len += 1 + max_item_len_len + max_item_len;
+
+
```

Comment:
random comment: in the final version it would be nice if there was not as much whitespace (e.g., here - just one empty line separating)